0000-zmix/README.md

```diff
@@ -0,0 +1,200 @@
+The proof format is described [here](proof.json)
+
+ZMix will provide objects for composing proofs directly in Rust or serializing
+from JSON.
+
+The following code is proposed
+
+```rust
+#[derive(Serialize, Deserialize)]
+pub enum ProofSpecClause {
+    Credential(...),
+    Interval(...),
+    SetMembership(...),
+    VerifiableEncryption(...),
+    Commitment(...),
+    ScopeCommitment(...)
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProofSpec {
+    pub clauses: HashMap<String, ProofSpecClause>
+}
+
+impl ProofSpec {
+    pub fn new() -> ProofSpec {
+        ProofSpec {
+            clauses: HashMap::new()
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum WitnessClause {
+    Credential(...),
+    Interval(...),
+    SetMembership(...),
+    VerifiableEncryption(...),
+    Commitment(...),
+    ScopeCommitment(...)
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Witness {
+    pub clauses: HashMap<String, WitnessClause>
+}
+
+impl Witness {
+    pub fn new() -> Witness {
+        Witness {
+            clauses: HashMap::new()
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SubProof {
+    ...
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Proof {
+    subproofs: HashMap<String, SubProof>
+}
+
+impl Proof {
+    pub fn create(proof_spec: &ProofSpec, witness: &Witness) -> Result<Proof, Error>;
+
+    pub fn verify(&self, proof_spec: &ProofSpec) -> Result<bool, Error>;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There is complexity that comes with making a generic zero knowledge language.
+Everytime a new primitive is added the language will need to be updated.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+There are quite a few ZKP libraries but they all are designed to focus
+on one specific kind of ZKP.
+[curve25519-dalek](https://doc.dalek.rs/zkp/index.html) uses Curve25519 and Ristretto
+ to make bulletproofs but is still considered experimental.
+[Bulletproofs](https://crypto.stanford.edu/bulletproofs/) have also been implemented using
+libsecp256k1 for BitCoin.
+[Libsnark](https://github.com/scipr-lab/libsnark) is for general purpose circuit based
+proofs but hasn't been updated in many years and is written in C++.
+ZCash has published and is using their [bellman](https://github.com/zkcrypto/bellman)
+library in production but its primary focus is again a single use case.
+[Idemix](https://idemix.wordpress.com) is an project based on anonymous credentials
+but is written in Java and not easily consumed.
+
+# Prior art
+[prior-art]: #prior-art
+
+[Camenisch-Stadler](http://soc1024.ece.illinois.edu/teaching/ece598am/fall2016/zkproofs.pdf) developed a notation for describing ZKP statements at a mathematical level but not at a code level.
+[Merlin](https://doc.dalek.rs/merlin/index.html) is a system of transcripts to describe ZKPs using this notation, but hasn't seen any other adoption than the dalek libraries and is also considered experimental.
```

Comment:
`Merlin` does not belong here as its objective is not to be a language for describing zero knowledge statements but only to a convenience tool to manage transcripts.
In my opinion, some competing projects are Snarky, Zokrates and Circom